module/vision/BallDetector/src/BallDetector.cpp

```diff
@@ -117,61 +118,79 @@ namespace module::vision {
+                balls->timestamp = horizon.timestamp;  // time when the image was taken
+                balls->Hcw       = horizon.Hcw;        // world to camera transform at the time the image was taken
 
+                // Check each cluster to see if it's a valid ball
                 for (auto& cluster : clusters) {
                     Ball b;
 
                     // Average the cluster to get the cones axis
                     Eigen::Vector3f axis = Eigen::Vector3f::Zero();
+                    // Add up all the unit vectors of each point (camera to point in world space) in the cluster
                     for (const auto& idx : cluster) {
                         axis += rays.col(idx);
                     }
-                    axis /= cluster.size();
+                    axis /= cluster.size();  // can this line be skipped if it's being normalised in the next line?
                     axis.normalize();
 
                     // Find the ray with the greatest distance from the axis
-                    // Should we use the average distance instead?
                     float radius = 1.0f;
                     for (const auto& idx : cluster) {
                         const Eigen::Vector3f& ray(rays.col(idx));
-                        if (axis.dot(ray) < radius) {
-                            radius = axis.dot(ray);
-                        }
+                        // arccos of the dot product gives the angle between the rays
+                        // the smaller arccos is, the bigger the angle - if between 0 and pi
+                        radius = axis.dot(ray) < radius ? axis.dot(ray) : radius;
                     }
 
-                    // Ball cam space info
-                    b.cone.axis   = horizon.Hcw.topLeftCorner<3, 3>().cast<float>() * axis;
-                    b.cone.radius = radius;
+                    // Set cone information for the ball
+                    // The rays are in world space, multiply by Rcw to get the axis in camera space
+                    b.cone.axis = horizon.Hcw.topLeftCorner<3, 3>().cast<float>()
+                                  * axis;    // Hcw is not an affine transform type (cannot use linear())
+                    b.cone.radius = radius;  // **arccos of the angle** between the furthest vectors
 
                     // https://en.wikipedia.org/wiki/Angular_diameter
+                    // From the link, the formula with arcsin is used since a ball is a spherical object
+                    // The variables are:
+                    //      delta: arccos(radius)
+                    //      d_act: field.ball_radius
+                    //      D: distance
+                    // Rearranging the equation gives
+                    //      distance = field.ball_radius / (2 * sin(arccos(radius)/2))
+                    // Using sin(arccos(x)) = sqrt(1 - x^2)
+                    //      distance = field.ball_radius / ( 2 * sqrt( 1 - (radius/2)^2 ) )
+                    // !!
```

Comment:
My calculation didn't give me what is in the code?

Reply:
The existing calculation seems to be okay when testing in Webots, but I'm not sure why. What I think it should be is almost twice the size.

Reply:
It's the radius vs diameter that you're having trouble with is my guess.
d_act is `2r` not `r`
Therefore in your rearranged math the 2s cancel giving what is in the code below

![image](https://user-images.githubusercontent.com/957962/172271247-17c274c7-0c52-4100-8e11-36c38596aa9c.png)

Reply:
Even with that, I don't get what is below, I get
$$ D = \frac{R}{(sin(\frac{arccos(r)}{2}))} $$

Doing it out in full:

$$ \delta = 2 \arcsin \left( \frac{d_{act}}{2D} \right) $$

Let $\delta = \arccos(r)$, where $r$ is `radius` in the code
Let $d_{act} = 2R$, where $R$ is `field.ball_radius`
and $D$ is the distance we're trying to find

$$ \arccos(r) = 2 \arcsin ( \frac{2R}{2D} ) $$

$$ \arccos(r) = 2 \arcsin ( \frac{R}{D} ) $$

$$ \frac{\arccos(r)}{2} = \arcsin ( \frac{R}{D} ) $$

$$ \sin \left( \frac{\arccos(r)}{2} \right) = \frac{R}{D} $$

$$ D = \frac{R}{\sin \left( \frac{\arccos(r)}{2} \right)} $$

Reply:
Ah also your second radius is off by 2 as well, The `radius` that you have is the dot product from the axis of the cone to a point on its edge, what the angular diameter of this equation wants is the entire diameter of the shape, this is twice the angle from the edge to the centre of the cone. So 2*acos(r) which cancels with your /2